widget/src/penrose.tsx

```diff
@@ -0,0 +1,271 @@
+    return digestArr.map(b => b.toString(16).padStart(2, '0')).join('')
+}
+
+/** The compile -> optimize -> prepare SVG sequence is not cheap (on the order of 1s for a simple diagram),
+ * so we aggressively cache its SVG outputs. */
+// TODO(WN): provide a "redraw" button to resample a misshapen diagram.
+const diagramSvgCache = new Map<string, SVGSVGElement>()
+
+function aliasToNumber (x: SVG.NumberAlias): number {
+    let y: string | number
+    if (x instanceof Number) y = x.valueOf()
+    else y = x as any
+
+    if (typeof y === 'string') return parseFloat(y)
+    else return y
+}
+
+async function renderPenroseTrio(trio: PenroseTrio, maxOptSteps: number): Promise<SVGSVGElement> {
+    console.log(trio.sty)
+    const hash = await hashTrio(trio)
+    if (diagramSvgCache.has(hash)) return diagramSvgCache.get(hash)!
+    const {dsl, sty, sub} = trio
+    const compileRes = penrose.compileTrio({ domain: dsl, style: sty, substance: sub, variation: '' })
+    if (compileRes.isErr()) throw new Error(penrose.showError(compileRes.error))
+    const state = await penrose.prepareState(compileRes.value)
+    const stateRes = penrose.stepUntilConvergence(state, maxOptSteps)
+    if (stateRes.isErr()) throw new Error(penrose.showError(stateRes.error))
+    if (!penrose.stateConverged(stateRes.value))
+        console.warn(`Diagram failed to converge in ${maxOptSteps} steps`)
+    const svg = await penrose.RenderStatic(stateRes.value, async path => path)
+
+    // The canvas is usually too large, so trim the SVG as a postprocessing step
+    const obj = SVG.SVG(svg)
+    const view = obj.viewbox()
+    let minX = view.width, maxX = 0, minY = view.height, maxY = 0
+
+    obj.each((i, children) => {
+        const child = children[i]
+        minX = Math.min(minX, aliasToNumber(child.x()))
+        maxX = Math.max(maxX, aliasToNumber(child.x()) + aliasToNumber(child.width()))
+        minY = Math.min(minY, aliasToNumber(child.y()))
+        maxY = Math.max(maxY, aliasToNumber(child.y()) + aliasToNumber(child.height()))
+    })
+
+    const padding = 10
+    const newX = minX - padding, newY = minY - padding, newW = (maxX - minX) + padding, newH = (maxY - minY) + padding
+    const newSvg = obj.viewbox(newX, newY, newW, newH).width(newW).height(newH)
+    diagramSvgCache.set(hash, newSvg.node)
+
+    return newSvg.node
+}
+
+export interface PenroseCanvasProps {
+    trio: PenroseTrio
+    maxOptSteps: number
+    embedNodes: Map<string, React.ReactNode>
+}
+
+interface InnerWithContainerProps extends PenroseCanvasProps {
+    containerDiv: HTMLDivElement
+    hiddenDiv: HTMLDivElement
+}
+
+interface InnerWithEmbedsProps extends Omit<InnerWithContainerProps, 'embedNodes'> {
+    embeds: Map<string, HTMLDivElement>
+}
+
+function InnerWithEmbeds({trio: {dsl, sty, sub}, maxOptSteps, embeds, containerDiv, hiddenDiv}: InnerWithEmbedsProps): JSX.Element {
+    const containerRect = containerDiv.getBoundingClientRect()
```

Comment:
Not sure if this kind of stuff is important at this stage but -- it's not great to read bounding rect during rendering. It's not deterministic and can also cause unnecessary layout or style recalculation. Ideally it would be snapshotted at some earlier time, or if we want to continuously track the container size, there should be some kind of resize observer to ensure the value stays fresh. Same for other getBoundingClientRect calls during render.

Apologies if this is a prototype / not intended for review, I just happened to walk by

Reply:
Thanks! I am guessing a resize observer which calls `setContainerRectDim` or some such piece of `React.state` would be better? If you have any pointers to documentation or articles about this specific use case, it would also be helpful.

Reply:
Yes, that’s the usual solution. Of course, if it’s possible to get rid of absolute pixel calculations and use percentages, that would be even better — but if it’s not, maybe something like https://github.com/ZeeCoder/use-resize-observer would work?

Reply:
I was able to remove most uses of `getBoundingClientRect` except for one which seems kind of hard to remove, but where can expect the size to be stable (it is the size of a programmatically generated element).